ostree-upload-server.py

```diff
@@ -251,16 +276,18 @@ def webapp_callback(task_name, filepath):
                         logging.debug("idle, do maintenance")
                         workers.stop()
 
-                        try:
-                            output = check_output(["flatpak",
-                                                   "build-update-repo",
-                                                   "--generate-static-deltas",
-                                                   "--prune",
-                                                   self._repo],
-                                                  stderr=STDOUT)
-                            logging.info("completed maintenance: " + output)
-                        except CalledProcessError as e:
-                            logging.error("failed maintenance: " + e.output)
+                        with RepoLock(self._repo, exclusive=True):
+                            try:
+                                output = check_output(["flatpak",
+                                                       "build-update-repo",
+                                                       "--generate-static-deltas",
+                                                       "--prune",
```

Comment:
I think that I would break it into a couple parts.

* I would do a simple summary update after each upload (ideally under exclusive lock). This should be pretty fast and ensures that you can actually see the new bundle when the upload completes.

* Generate deltas and/or prune when idle. As mentioned, these can be very heavy operations, so I don't think we want to do them often. The other part about the delta generation is that we have a separate delta policy that's more involved than the simple builtin flatpak policy. It involves looking at the released/production refs and trying to make deltas to those. So, we'll probably have some external process that does the delta heavy lifting. Same with the pruning. So, it might be nice to make it possible to not run these maintenance tasks or possibly providing custom maintenance tasks.